src/gadgets/curves/zeroable_affine/mod.rs

```diff
@@ -1,73 +1,277 @@
-    pub y: NN,
-    pub is_zero: Boolean<F>,
-    pub _marker: std::marker::PhantomData<C>,
+    x: NF,
+    y: NF,
+    pub is_infinity: Boolean<F>,
+    pub _marker: PhantomData<GC>,
 }
 
-// we only need add/sub/double/negate Mul is implemented by naive double-and-add, and we can have special
-// mul that will multiply by an element of scalar field, where zeroness-exception can only happen once.
-
-// We also create decompress function for convenience
-
-impl<F: SmallField, C: GenericCurveAffine, NN: NonNativeField<F, C::Base>>
-    ZeroableAffinePoint<F, C, NN>
+impl<F, GC, NF> ZeroableAffinePoint<F, GC, NF>
 where
-    C::Base: pairing::ff::PrimeField,
+    F: SmallField,
+    GC: GenericCurveAffine,
+    NF: NonNativeField<F, GC::Base>,
+    GC::Base: PrimeField,
 {
-    pub fn zero_point<CS: ConstraintSystem<F>>(cs: &mut CS, params: &Arc<NN::Params>) -> Self {
-        use pairing::ff::Field;
-        let zero_nn = NN::allocated_constant(cs, C::Base::zero(), params);
-        let boolean_true = Boolean::allocated_constant(cs, true);
+    /// Initializes a new non-infinite affine point with the specified coordinates
+    fn new<CS>(cs: &mut CS, x: NF, y: NF) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        Self {
+            x,
+            y,
+            is_infinity: Boolean::allocated_constant(cs, false),
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns the x-coordinate of the point
+    fn x(&self) -> &NF {
+        &self.x
+    }
+
+    /// Returns the y-coordinate of the point
+    fn y(&self) -> &NF {
+        &self.y
+    }
+
+    /// Initializes the point at infinity. x and y are set to zero, and is_infinity is set to true.
+    fn zero_point<CS>(cs: &mut CS, params: &Arc<NF::Params>) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let zero_nf = NF::allocated_constant(cs, GC::Base::zero(), params);
 
         Self {
-            x: zero_nn.clone(),
-            y: zero_nn,
-            is_zero: boolean_true,
-            _marker: std::marker::PhantomData,
+            x: zero_nf.clone(),
+            y: zero_nf,
+            is_infinity: Boolean::allocated_constant(cs, true),
+            _marker: PhantomData,
         }
     }
 
-    pub fn same_x<CS: ConstraintSystem<F>>(&mut self, cs: &mut CS, other: &mut Self) -> Boolean<F> {
+    /// Adds two affine points on the curve.
+    fn add<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // If x's are not the same, adding via unequal_x method
+        let same_x = self.same_x(cs, other);
+        let boolean_false = Boolean::allocated_constant(cs, false);
+        if same_x == boolean_false {
+            return self.add_unequal_x(cs, other);
+        }
+
+        // If y's are the same, doubling the point
+        let same_y = self.same_y(cs, other).negated(cs);
+        if same_y == boolean_false {
+            return self.double(cs);
+        }
```

Comment:
also non-homogenous, should be done with conditional select gates

Reply:
Just removed the function as of now; we probably will not need it :) But we can safely add it further